### DISCUSSION TITLE
Combine streams together

### DISCUSSION THREAD

**Discussion by T1j3ff** [Motion-Project/motion#1717] [Q&A]
As I understand, Motion outputs MJPEG streams 

Is there a way to 4 streams together to form only one stream ?
So we can see all 4 images in a single stream 

With motion , or something else ?

**Comment by genius3000** [comment]
You can't do this with motion itself. VLC does have a [mosaic module](https://wiki.videolan.org/VLC_HowTo/Make_a_mosaic/). I've toyed with it briefly and could set up multiple MJPEG streams into one view. You can change the output methods, pretty sure you can set it for a stream. My main goal was chromecast output which I've had limited success with so far.

**Comment by T1j3ff** [comment]
Thanks alot for the cue... 
For now I'm tiring ffmpeg for video processing, and LiveGo to re-stream . I'll keep you posted

**Comment by T1j3ff** [comment]
Its a bit convoluted , but I got this to work  

To prevent image degradations and delay (to many recompressions) I'm taking the stream from the cameras themself 
so this doesn't involve Motion Mjpeg anymore, 
But this could interest people here i'm shure anyway 

The beauty of this , is that cameras qty and size it is only limited by CPU power 
![image](https://github.com/Motion-Project/motion/assets/11382177/3c397a02-6ed5-410e-82d9-5b0376cecf72)

I'm probably making many thing wrong (specialty in the ffmpeg re-encoding/streaming), but this works for me 
![image](https://github.com/Motion-Project/motion/assets/11382177/fbf29415-2cd5-4ad3-97a6-c6882cca3976)

Info: 
Installing liveGo (Skip ffmpeg instalation to "Install LiveGo")
https://medium.com/an-idea/there-can-be-only-one-one-camera-multiple-streams-7a1142a35905
Processing the cameras 
https://trac.ffmpeg.org/wiki/Create%20a%20mosaic%20out%20of%20several%20input%20videos

Smash together Final FFMPEG command (all on one line) 

ffmpeg 
-rtsp_transport tcp -an -i "rtsp://User:Pass@192.168.1.101:554" 
-rtsp_transport tcp -an -i "rtsp://User:Pass@192.168.1.102:554" 
-rtsp_transport tcp -an -i "rtsp://User:Pass@192.168.1.103:554" 
-rtsp_transport tcp -an -i "rtsp://User:Pass@192.168.1.104:554" 
-filter_complex "
nullsrc=size=1920x1080 [base]; 
[0:v] setpts=PTS-STARTPTS, scale=940x520 [upperleft]; 
[1:v] setpts=PTS-STARTPTS, scale=940x520 [upperright]; 
[2:v] setpts=PTS-STARTPTS, scale=940x520 [lowerleft]; 
[3:v] setpts=PTS-STARTPTS, scale=940x520 [lowerright]; 
[base][upperleft] overlay=shortest=1:x=10:y=10 [tmp1]; 
[tmp1][upperright] overlay=shortest=1:x=960:y=10 [tmp2]; 
[tmp2][lowerleft] overlay=shortest=1:x=10:y=540 [tmp3]; 
[tmp3][lowerright] overlay=shortest=1:x=960:y=540" 
-c:v libx264 -preset ultrafast -tune zerolatency -b:v 2000k -bsf:v h264_mp4toannexb -maxrate 3000k -bufsize 2000k -f flv rtmp://127.0.0.1:1935/live/***********Livegotoken*******************

**Reply by genius3000** [reply]
Thanks for the update, and the detailed solution notes. I'm going to toy around with the ffmpeg mosaic and see if it works better for my usage.

**Comment by T1j3ff** [comment]
I'm not a pro in any of these things 

If anybody know how to remove LiveGo from the loop this would remove some complexity too 
(but honnestly , I have nothing to complaint on it for now ,, it work great, havent crashed , and is not cpu intensive) 


But I had many issue with FFMPEG 

First I had  :  " Thread message queue blocking; consider raising the thread_queue_size option (current value: 8) " 
Raising to 256 seems the sweet spot for me to remove (most) of these . Basicaly ffmpeg wasnt  able to process frames that fast and Had to drop some ,,, raising this (buffer?) helped 

After I got many of these  "RTP: PT=60: bad cseq 45fc expected=37ddr -"
This is related to ffmpeg receiving packets from a camera out of order 
I've pin point this to only one of my 4 cams 
The only one in 8k 
I've reduce the resolution , and bitrate a bit , and most of these error are gone 

"[libx264 @ 0x5594ff9e4080] VBV underflow (frame 364001, -20992 bits)=1997.7kbits/s speed=   1x" 
Fidling with the encoding outputs buffer resolved these 

"-an" option remove audio, so it remove audio related errors 


This is my last ffmpeg recipee (all on one line) (Ymmv)
 

ffmpeg 
-thread_queue_size 256 -rtsp_transport tcp -an -i "rtsp://User:Pass@192.168.1.101:554" 
-thread_queue_size 256 -rtsp_transport tcp -an -i "rtsp://User:Pass@192.168.1.102:554" 
-thread_queue_size 256 -rtsp_transport tcp -an -i "rtsp://User:Pass@192.168.1.103:554" 
-thread_queue_size 256 -rtsp_transport tcp -an -i "rtsp://User:Pass@192.168.1.104:554" 
-filter_complex "
nullsrc=size=1920x1080 [base]; 
[0:v] setpts=PTS-STARTPTS, scale=940x520 [upperleft]; 
[1:v] setpts=PTS-STARTPTS, scale=940x520 [upperright]; 
[2:v] setpts=PTS-STARTPTS, scale=940x520 [lowerleft]; 
[3:v] setpts=PTS-STARTPTS, scale=940x520 [lowerright]; 
[base][upperleft] overlay=shortest=1:x=10:y=10 [tmp1]; 
[tmp1][upperright] overlay=shortest=1:x=960:y=10 [tmp2]; 
[tmp2][lowerleft] overlay=shortest=1:x=10:y=540 [tmp3]; 
[tmp3][lowerright] overlay=shortest=1:x=960:y=540" 
-c:v libx264 -preset ultrafast -tune zerolatency -b:v 2000k -bsf:v h264_mp4toannexb -maxrate 3000k -bufsize 4000k -f flv rtmp://127.0.0.1:1935/live/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

This one still crashed FFMPEG after 12h :( 
Also there is a significant processing delay between cams .... 10-30 sec 
To be continued  

Also, this use ~25% of my  i7-6700 ,,, 24-7,,,  (200% for 8cores .. in top)
